.github/workflows/main.yml

```
@@ -0,0 +1,261 @@
+    name: Spellcheck
+    runs-on: ubuntu-latest
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: solana-program/actions/setup-ubuntu@main
+        with:
+          default-toolchain: true
+          cli: true # needed for llvm-config
+          cargo-cache-key: cargo-spellcheck
+
+      - name: Install cargo-spellcheck
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-spellcheck
+
+      - name: Run cargo-spellcheck
+        run: make spellcheck
+
+  format_and_lint_rust:
+    name: Format & Lint Rust
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        package: ${{ fromJson(inputs.rust-packages) }}
+      fail-fast: false
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: solana-program/actions/setup-ubuntu@main
+        with:
+          clippy-toolchain: ${{ inputs.clippy-toolchain }}
+          rustfmt-toolchain: ${{ inputs.rustfmt-toolchain }}
+          cli: true
+          cargo-cache-key: cargo-lint-${{ matrix.package }}
+
+      - name: Install cargo-hack
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-hack
+
+      - name: Format
+        run: make format-check-${{ matrix.package }}
+
+      - name: Clippy
+        run: make clippy-${{ matrix.package }}
+
+      - name: Lint / Docs
+        run: make build-doc-${{ matrix.package }}
+
+      - name: Check powerset
+        run: make powerset-${{ matrix.package }}
+
+  wasm:
+    name: Build WASM
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        package: ${{ fromJson(inputs.wasm-packages) }}
+      fail-fast: false
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: solana-program/actions/setup-ubuntu@main
+        with:
+          default-toolchain: true
+          cargo-cache-key: cargo-wasm-${{ matrix.package }}
+
+      - name: Install wasm target
+        run: rustup target add wasm32-unknown-unknown
+
+      - name: Build with wasm target
+        run: make build-wasm-${{ matrix.package }}
+
+  build_sbpf:
+    name: Build Program
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        package: ${{ fromJson(inputs.sbpf-program-packages) }}
+      fail-fast: false
+    steps:
+      - name: Git Checkout
+        uses: actions/checkout@v4
+
+      - name: Setup Environment
+        uses: solana-program/actions/setup-ubuntu@main
+        with:
+          default-toolchain: true
+          cargo-cache-key: cargo-build-sbf-${{ matrix.package }}
+          solana: ${{ inputs.solana-cli-version }}
+
+      - name: Build
+        run: make build-sbf-${{ matrix.package }}
+
+      - name: Save Program Builds For Client Jobs
+        uses: actions/cache/save@v4
+        with:
+          path: ./target/deploy/*.so
+          key: ${{ runner.os }}-builds-${{ github.sha }}
```

nit: `builds` is the only static part of the key here and it seems pretty generic considering we are building all sorts of things within this multi-language repo. Maybe something like `program-builds` would be better?

```suggestion
          key: ${{ runner.os }}-program-builds-${{ github.sha }}
```

---

You got it! https://github.com/solana-program/actions/pull/4/commits/57c522a2194b47ccf0cc530dd0591035923e4e82